sync/azuredatabricks.py

```diff
@@ -0,0 +1,411 @@
+            vms = compute.virtual_machines.list(resource_group_name=resource_group_name)
+        else:
+            logger.warning("Failed to find Databricks managed resource group")
+            vms = compute.virtual_machines.list_all()
+
+        cluster_instances = [
+            vm.as_dict()
+            for vm in vms
+            if vm.tags.get("Vendor") == "Databricks"
+            and vm.tags.get("ClusterId") == cluster["cluster_id"]
+        ]
+
+    if not cluster_instances:
+        no_instances_message = (
+            f"Unable to find any active or recently terminated instances for cluster `{cluster_id}`. "
+            + "Please refer to the following documentation for options on how to address this - "
+            + "https://docs.synccomputing.com/sync-gradient/integrating-with-gradient/databricks-workflows"
+        )
+        return Response(error=DatabricksError(message=no_instances_message))
+
+    return Response(result=cluster_instances)
+
+
+def monitor_cluster(cluster_id: str, polling_period: int = 30) -> None:
+    cluster = get_default_client().get_cluster(cluster_id)
+    spark_context_id = cluster.get("spark_context_id")
+
+    while not spark_context_id:
+        # This is largely just a convenience for when this command is run by someone locally
+        logger.info("Waiting for cluster startup...")
+        sleep(30)
+        cluster = get_default_client().get_cluster(cluster_id)
+        spark_context_id = cluster.get("spark_context_id")
+
+    (log_url, filesystem, bucket, base_prefix) = _cluster_log_destination(cluster)
+    if log_url:
+        _monitor_cluster(
+            (log_url, filesystem, bucket, base_prefix), cluster_id, spark_context_id, polling_period
+        )
+    else:
+        logger.warning("Unable to monitor cluster due to missing cluster log destination - exiting")
+
+
+def _monitor_cluster(
+    cluster_log_destination, cluster_id: str, spark_context_id: int, polling_period: int
+) -> None:
+    (log_url, filesystem, bucket, base_prefix) = cluster_log_destination
+    # If the event log destination is just a *bucket* without any sub-path, then we don't want to include
+    #  a leading `/` in our Prefix (which will make it so that we never actually find the event log), so
+    #  we make sure to re-strip our final Prefix
+    file_key = f"{base_prefix}/sync_data/{spark_context_id}/cluster_instances.json".strip("/")
+
+    if filesystem == "dbfs":
+        path = format_dbfs_filepath(file_key)
+        dbx_client = get_default_client()
+
+        def write_file(body: bytes):
+            logger.info("Saving state to DBFS")
+            write_dbfs_file(path, body, dbx_client)
+
+    resource_group_name = _get_databricks_resource_group_name()
+    if not resource_group_name:
+        logger.warning("Failed to find Databricks managed resource group")
+
+    previous_instances = {}
+
+    while True:
+        try:
+            compute = _get_azure_client(ComputeManagementClient)
```

Comment:
nit - any reason for this to be inside the `while`?

Reply:
Nope - I had meant to move that up.